### ISSUE TITLE
Larger output with -o x.wasm as opposed to -s STANDALONE_WASM

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#14543]
Targetting a wasm file turns on standalone mode, but it also does more things it seems, that increase code size.

For example,
```
./emcc tests/other/metadce/minimal.c -O3 -o a.out.js --no-entry -s STANDALONE_WASM
```
emits 124 bytes, while replacing `.js` with `.wasm` leads to `266` bytes. (Can also remove `-s STANDALONE_WASM` in that mode, it doesn't matter.)

I can look into this, but I wondered if that has an obvious explanation perhaps? cc @sbc100

**Comment by aduros** [comment]
Thanks for opening this.

I found your other article about compiling the add.c example into a 42 byte (!!) .wasm https://hacks.mozilla.org/2018/01/shrinking-webassembly-and-javascript-code-sizes-in-emscripten/ I'd love if this was still possible!

**Comment by sbc100** [comment]
I would guess this is something to do with metadce?   When the compiler can can see the JS that will load the wasm perhaps assumes that any exports not needed by the JS can be removed.   Perhaps this is a bad assumption and `STANDALONE_WASM` should never do DCE (that would make sense since meta-DCE is fundamentally makes the wasm specific the JS loading code).

**Comment by sbc100** [comment]
When we include JS in the output the exports get reduced from:

```
 - memory[0] -> "memory"
 - func[1] <add> -> "add"
 - global[1] -> "global_val"
 - table[0] -> "__indirect_function_table"
 - func[0] <__wasm_call_ctors> -> "_initialize"
 - func[5] <__errno_location> -> "__errno_location"
 - func[2] <stackSave> -> "stackSave"
 - func[3] <stackRestore> -> "stackRestore"
 - func[4] <stackAlloc> -> "stackAlloc"
```

```
 - memory[0] -> "memory"
 - func[1] <add> -> "add"
 - table[0] -> "__indirect_function_table"
 - func[0] <__wasm_call_ctors> -> "_initialize"
```

The reason is that we call `building.minify_wasm_js` in one case and not the other.

Should we always avoid calling `building.minify_wasm_js` in `STANDALONE_WASM` mode since calling that function essentially specializes the wasm to the JS which makes the wasm less standalone (I think)?

**Comment by kripken** [comment]
Looking into this, yes, the core issue is that we don't run `building.minify_wasm_js`. But we can't run it in standalone mode when only emitting wasm, as then there is no JS to run with.

I believe we used to emit JS all the way to the end, and then remove it at that point. But some emcc.py refactoring must have changed that, and silently regressed this. One option might be to go back to that approach. @sbc100 maybe you remember where this happened, and how easy it would be to partially revert?

Another option might be to write a stripped-down metadce just for wasm-only mode, "faking" the data we normally get from the JS.

Another option might be to refactor how we handle exports. The problem is that the emcc adds exports like `stackSave`, which metadce then removes if it isn't actually used by JS. In standalone mode, we don't actually want to ever export such a function, as it would not work in any wasm VM, it only works with the custom JS. So we could refactor emcc.py to change all the places where we do `EXPORTED_FUNCTIONS += ["foo"]` to annotate those functions as either

1. A function exported for JS. We don't need this in wasm-only mode.
2. A function exported because we want to link in that code (which might be used even if we don't call it, as it might bring in a global constructor).

With that in place, instead of metadce we could just tell wasm-ld to only preserve the user-provided exports plus exports of type 2 and not type 1. We would also need to audit all the JS emitting code to avoid using exports of type 1 (e.g., in my experiment, this breaks on `emscripten_init_stack` - something we indeed do not want in standalone mode, but currently emit; we'd need to modify how we emit JS for such things).

What do you think @sbc100 ?

**Comment by sbc100** [comment]
I think the solution is to avoid running `minify_wasm_js` in `STANDALONE_WASM`.   In `STANDALONE_WASM` we do *not* want to specialize to the JS since the wasm file might be used elsewhere in some other environment.      It should be standalone. 

I don't think we want to be removing any exports.  Even `stackSave` could well be needed by other embeddinger.. we can't/don't know.   Another use for standalone mode is that we might want to pass the wasm file a input to `emcc --post-link`.. and at the point the JS might require `stackSave`, etc,

**Comment by sbc100** [comment]
The side effect of course is that STANDALONE_WASM builds will always be slightly larger than embedded/fused builds.. but I think that is an inevitable part of being standalone.

**Comment by kripken** [comment]
I agree that in standalone mode we do not want to specialize to the JS. That's a good way to look at it. But the more I think about it, the less sure I am what that means. For example, we add `stackSave` to the wasm, and keep it alive, only because of specific needs of our current JS. Those needs change over time. So to be truly standalone, the wasm should not care about the needs of our current JS.

That is, running `minify_wasm_js / metadce` means to specialize the wasm to the JS. But also to even add `stackSave` as an export in the wasm, based on the JS's needs, is also to specialize the wasm to the JS. Ideally we would do neither, I think, which means we'd end up with a wasm that has the exports the user asked for, **and nothing more**.

Does that make sense?

**Comment by sbc100** [comment]
I see it this way:  Emscripten exports the stack manipulation functions in case the embedder, whatever it maybe be, might need to manipulate the stack.    You can imagine a custom embedder that is not emscripten could also need to use those function as part of implementing other functions (like emscripten sometimes does).   In STANDALONE_WASM mode we don't know what the embedder is or how they implement their library function.. we don't know if they need the stack manipulation functions. 

To an embedder (such as the emscripten JS library code) those stack functions are part of the ABI and we don't know if the embedder needs those ABI function until we specialize to a given embedder (i.e. we run `minify_wasm_js / metadce`).  

We can't eliminate them and we should assume the embedder needs access to them to help implement their embedding.

**Comment by kripken** [comment]
Maybe the question then is do we want to support such custom embeddings. I'd say yes for standard APIs like OpenGL, WASI, etc., but no for nonstandard internals of Emscripten itself.

My reasoning is that in standalone wasm mode we are emitting a wasm file that should run in wasm VMs as much as possible, with no JS support code. Emitting internal Emscripten details would reduce the chance of that working, especially since that ABI is not stable.

**Comment by sbc100** [comment]
Remember are talking about exports, no imports here.  I'm not sure exports can break any VMs.

I think it should be goal to be able to take files built with STANDALONE_WASM and then run them through `emcc --post-link` and have them work.    If we strip out the stack functions (or other functions needed by out JS code) we limit out ability to do that.

IIRC, `STANDALONE_WASM` is a best effort to hide implementation details.. we do it when we can but there are times when we can't and these stack manipulation functions are a necessary detail in some cases.

**Comment by sbc100** [comment]
Our standalone wasm files already *import* many non-standard things (for all but the most trivial programs)... this is way more harmfull that exporting a few non-standard things.

IMHO we should focus on removing non-standard imports (since they will certainly block running on standard VMs) before we looks at non-standard exports (which are mostly harmless).
